poseidon2/src/lib.rs

```diff
@@ -0,0 +1,289 @@
+            t7 += t4;
+
+            state[idx] = t6;
+            state[idx + 1] = t5;
+            state[idx + 2] = t7;
+            state[idx + 3] = t4;
+        }
+    }
+
+    fn matmul_external(&self, state: &mut [F; WIDTH]) {
+        match WIDTH {
+            2 => {
+                let sum = state[0] + state[1];
+                state[0] += sum;
+                state[1] += sum;
+            }
+            3 => {
+                let sum = state[0] + state[1] + state[2];
+                state[0] += sum;
+                state[1] += sum;
+                state[2] += sum;
+            }
+            4 => {
+                self.matmul_m4(state);
+            }
+            8 | 12 | 16 | 20 | 24 => {
+                self.matmul_m4(state);
+
+                let t = WIDTH;
+                let t4 = t / 4;
+                let mut stored = [F::ZERO; 4];
+                for l in 0..4 {
+                    stored[l] = state[l];
+                    for j in 1..t4 {
+                        stored[l] += state[4 * j + l];
+                    }
+                }
+                for i in 0..WIDTH {
+                    state[i] += stored[i % 4];
+                }
+            }
+            _ => panic!("WIDTH not supported"),
+        }
+    }
+
+    fn matmul_internal(&self, state: &mut [F; WIDTH], mat_internal_diag_m_1: &[F]) {
+        match WIDTH {
+            2 => {
+                let mut sum = state[0];
+                sum += state[1];
+                state[0] += sum;
+                state[1] += state[1];
+                state[1] += sum;
+            }
+            3 => {
+                let mut sum = state[0];
+                sum += state[1];
+                sum += state[2];
+                state[0] += sum;
+                state[1] += sum;
+                state[2] += state[2];
+                state[2] += sum;
+            }
+            4 | 8 | 12 | 16 | 20 | 24 => {
+                let mut sum = state[0];
+                for i in 1..WIDTH {
+                    sum += state[i];
+                }
+                for i in 0..WIDTH {
+                    state[i] *= mat_internal_diag_m_1[i];
+                    state[i] += sum;
+                }
+            }
+            _ => panic!("WIDTH not supported"),
+        }
+    }
+}
+
+impl<F, Mds, const WIDTH: usize, const D: usize> CryptographicPermutation<[F; WIDTH]>
+    for Poseidon2<F, Mds, WIDTH, D>
+where
+    F: Field,
+    Mds: MdsPermutation<F, WIDTH>,
+{
+    fn permute(&self, state: [F; WIDTH]) -> [F; WIDTH] {
+        let mut state = state.to_owned();
+
+        // self.mds.permute_mut(&mut state);
+        self.matmul_external(&mut state);
+
+        let rounds = self.rounds_f + self.rounds_p;
+        let rounds_f_beggining = self.rounds_f / 2;
+        for r in 0..rounds_f_beggining {
+            state = self.add_rc(&state, &self.round_constants[r]);
+            state = self.sbox(&state);
+            // self.mds.permute_mut(&mut state);
```

Comment:
Should I be calling `self.mds.permute_mut` everywhere here?

Reply:
Yeah I think so, so it can be generic w.r.t. the MDS transform used